vendor: add local copy of archive/tar

tiborvass opened the pull request:
This version avoids doing name lookups on creating tarball that
should be avoided in to not hit loading glibc shared libraries.

Signed-off-by: Tonis Tiigi <tonistiigi@gmail.com>
(cherry picked from commit aa6a9891b09cce3d9004121294301a30d45d998d)
Signed-off-by: Tibor Vass <tibor@docker.com>

Couple of changes from the commit that I cherry-picked:
- I completed the instructions at the end of vendor.conf (PTAL)
- I reran them inside the dev container (which uses go 1.13.8).

tonistiigi committed:
vendor: add local copy of archive/tar

This version avoids doing name lookups on creating tarball that
should be avoided in to not hit loading glibc shared libraries.

Signed-off-by: Tonis Tiigi <tonistiigi@gmail.com>
(cherry picked from commit aa6a9891b09cce3d9004121294301a30d45d998d)
Signed-off-by: Tibor Vass <tibor@docker.com>

tianon commented on patches/0001-archive-tar-do-not-populate-user-group-names.patch:
```diff
@@ -0,0 +1,79 @@
+
+This is ultimately fixed in Go 1.11 (provided that "osusergo" build tag
+is set (see https://github.com/golang/go/commit/62f0127d81). In the
+meantime, let's use this fork.
```
I'm guessing this description isn't accurate given we're patching `archive/tar` again in a post-1.11 Go version?

thaJeztah replied:
We reused the same commit that we used in the past (but at the time for a different reason)

tianon replied:
Right, that makes sense, but shouldn't we update the description?  From a naive reading it looks like this patch existing again is a mistake, since it's "fixed" in 1.11+. :sweat_smile:

Is there an issue filed on https://github.com/golang/go we could link to for further discussion?  Are they open to a knob for disabling this behavior in a future Go version that could re-deprecate this patch in the future?

thaJeztah replied:
@kolyshkin @tonistiigi ^^

tonistiigi replied:
There is no issue for making this behavior configurable afaik

kolyshkin replied:
No, AFAIK no one filed it. Makes sense to do so (at least to start a conversation). I'm pretty busy atm though :(

thaJeztah replied:
@tonistiigi could you open a ticket? You're probably better at describing the problem than I am

thaJeztah commented:
This fixes https://github.com/moby/moby/issues/40211 and https://github.com/moby/moby/issues/39449 on master

thaJeztah commented:
@tonistiigi or @kolyshkin could one of you look at https://github.com/moby/moby/pull/40672#discussion_r391853532 ?

@tianon would you like to have the patch description updated as part of this PR (probably an extra commit, because this was cherry-picked from the release branch), or in a follow-up?

tianon commented:
Well, for context, I've already had two people see this patch description in the diff for the security update and get confused enough about whether it applies (given the description implies the problem the patch is fixing is already fixed in newer Go releases) that they reached out and asked me about it privately, so IMO the description really should be updated before this merges at the very least to make it clear it's an old patch being used to fix a *new* issue and is not in fact fixed by newer Go versions. :sweat_smile:

tonistiigi commented:
Yes, this doesn't patch go but switches tar package from stdlib to our fork. Because vndr does not support stdlib packages natively some hackery is needed. An earlier version of this just used my fork branch like we do with some other vendors, but for better visibility of the changes, it was switched to using patch files.

AkihiroSuda commented:
What's current status?

AkihiroSuda commented:
needs rebase

tianon commented:
IMO the description in the patch file should either be updated or simply removed since it no longer applies to our usage, but otherwise I'm fine with this. :+1:

cpuguy83 committed:
Remove blurb about fixed in go1.11 in patch file

This blurb exists because we reused the same commit from an old patch,
and thus got the commit message with it. However the message is
confusing in this context.
It was suggested in review that we should remove the confusing message.

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

cpuguy83 commented:
This is rebased and the g1.11 message was removed (in a new commit).

tianon commented on patches/0001-archive-tar-do-not-populate-user-group-names.patch:
```diff
@@ -11,10 +11,6 @@ on NULL pointer dereference).
 
 For much more details, see https://github.com/golang/go/issues/23265
```
This link is also misleading, but at least it's not quite as confusing. :+1: :heart:

tianon reviewed (approved):
LGTM :heart:

AkihiroSuda commented:
Ready to merge?

thaJeztah reviewed (approved):
LGTM
